File: code/python/tree/rbt.py
class RBT:
    """A red-black tree."""

    RED = False
    BLACK = True

    class Node:
        """Node entry for RBT."""

        def __init__(self, key=None):
            self.key = key
            self.color = RBT.BLACK
            self.left = None
            self.right = None
            self.p = None

    def __init__(self):
        self._NIL = RBT.Node()
        self._root = self._NIL
        self._size = 0

    def size(self):
        return self._size

    def is_empty(self):
        return self._size == 0

    def get(self, key):
        x = self._root
        while x is not self._NIL and key != x.key:
            if key < x.key:
                x = x.left
            else:
                x = x.right
        return x

    def contains(self, key):
        if key is None:
            raise KeyError
        return self.get(key) is not self._NIL

    def _left_rotate(self, x: Node):
        y = x.right
        x.right = y.left
        if y.left is not self._NIL:
            y.left.p = x
        y.p = x.p
        if x.p is self._NIL:
            self._root = y
        elif x is x.p.left:
            x.p.left = y
        else:
            x.p.right = y
        y.left = x
        x.p = y

    def _right_rotate(self, y: Node):
        x = y.left
        y.left = x.right
        if x.right is not self._NIL:
            x.right.p = y
        x.p = y.p
        if y.p is self._NIL:
            self._root = x
        elif y is y.p.left:
            y.p.left = x
        else:
            y.p.right = x
        x.right = y
        y.p = x

    def put(self, key):
        x = self._root
        y = self._NIL
        while x is not self._NIL:
            y = x
            if key < x.key:
                x = x.left
            elif key > x.key:
                x = x.right
            else:
                return
        z = RBT.Node(key)
        z.p = y
        if y is self._NIL:
            self._root = z
        elif z.key < y.key:
            y.left = z
        else:
            y.right = z
        z.left = self._NIL
        z.right = self._NIL
        z.color = RBT.RED
        self._size += 1
        self._insert_fixup(z)

    def _insert_fixup(self, z: Node):
        while z.p.color == RBT.RED:
            if z.p is z.p.p.left:
                y = z.p.p.right
                if y.color == RBT.RED:
                    z.p.color = RBT.BLACK
                    y.color = RBT.BLACK
                    z.p.p.color = RBT.RED
                    z = z.p.p
                else:
                    if z is z.p.right:
                        z = z.p
                        self._left_rotate(z)
                    z.p.color = RBT.BLACK
                    z.p.p.color = RBT.RED
                    self._right_rotate(z.p.p)
            else:
                y = z.p.p.left
                if y.color == RBT.RED:
                    z.p.color = RBT.BLACK
                    y.color = RBT.BLACK
                    z.p.p.color = RBT.RED
                    z = z.p.p
                else:
                    if z is z.p.left:
                        z = z.p
                        self._right_rotate(z)
                    z.p.color = RBT.BLACK
                    z.p.p.color = RBT.RED
                    self._left_rotate(z.p.p)
        self._root.color = RBT.BLACK

    def _min(self, x: Node):
        while x.left is not self._NIL:
            x = x.left
        return x

    def _transplant(self, u: Node, v: Node):
        if u.p is self._NIL:
            self._root = v
        elif u is u.p.left:
            u.p.left = v
        else:
            u.p.right = v
        v.p = u.p

    def _delete(self, z: Node):
        if z.left is not self._NIL and z.right is not self._NIL:
            y = self._min(z.right)
            z.key = y.key
            z = y
        x = z.left if z.left is not self._NIL else z.right
        self._transplant(z, x)
        self._delete_fixup(x)

    def _delete_fixup(self, x: Node):
        while x is not self._root and x.color == RBT.BLACK:
            if x is x.p.left:
                w = x.p.right
                if w.color == RBT.RED:
                    w.color = RBT.BLACK
                    x.p.color = RBT.RED
                    self._left_rotate(x.p)
                    w = x.p.right
                if w.left.color == RBT.BLACK and w.right.color == RBT.BLACK:
                    w.color = RBT.RED
                    x = x.p
                else:
                    if x.right.color == RBT.BLACK:
                        x.left.color = RBT.BLACK
                        w.color = RBT.RED
                        self._right_rotate(w)
                        w = x.p.right
                    w.color = x.p.color
                    x.p.color = RBT.BLACK
                    w.right.color = RBT.BLACK
                    self._left_rotate(x.p)
                    x = self._root
            else:
                w = x.p.left
                if w.color == RBT.RED:
                    w.color = RBT.BLACK
                    x.p.color = RBT.RED
                    self._right_rotate(x.p)
                    w = x.p.left
                if w.right.color == RBT.BLACK and w.left.color == RBT.BLACK:
                    w.color = RBT.RED
                    x = x.p
                else:
                    if w.left.color == RBT.BLACK:
                        w.right.color = RBT.BLACK
                        w.color = RBT.RED
                        self._left_rotate(w)
                        w = x.p.left
                    w.color = x.p.color
                    x.p.color = RBT.BLACK
                    x.left.color = RBT.BLACK
                    self._right_rotate(x.p)
                    x = self._root
        x.color = RBT.BLACK

    def remove(self, key):
        if key is None:
            raise KeyError
        z = self.get(key)
        if z is not self._NIL:
            self._delete(z)
            self._size -= 1

    def height(self):
        def _height(x: RBT.Node):
            if x is self._NIL:
                return -1
            else:
                return 1 + max(_height(x.left), _height(x.right))
        return _height(self._root)

    def in_order(self):
        def _in_order(x: RBT.Node):
            if x is not self._NIL:
                _in_order(x.left)
                print(x.key)
                _in_order(x.right)
        _in_order(self._root)

File: code/java/tree/src/main/java/org/swufe/RBT.java
package org.swufe;
/**
 * A red-black tree.
 */
public class RBT<Key extends Comparable<Key>> {

    private static final boolean RED   = false;
    private static final boolean BLACK = true;

    private static class Node<Key extends Comparable<Key>> {
        Key key;
        boolean color;
        Node<Key> left;
        Node<Key> right;
        Node<Key> p;

        public Node(Key key) {
            this.key = key;
            color = RED;
            left = null;
            right = null;
            p = null;
        }
    }

    private Node<Key> root;
    private int size;

    public RBT() {
        root = null;
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    private void rotateLeft(Node<Key> x) {
        assert x.right != null;
        Node<Key> y = x.right;
        x.right = y.left;
        if (y.left != null) {
            y.left.p = x;
        }
        y.p = x.p;
        if (x.p == null) {
            root = y;
        } else if (x == x.p.left) {
            x.p.left = y;
        } else {
            x.p.right = y;
        }
        y.left = x;
        x.p = y;
    }

    private void rotateRight(Node<Key> y) {
        assert y.left != null;
        Node<Key> x = y.left;
        y.left = x.right;
        if (x.right != null) {
            x.right.p = y;
        }
        x.p = y.p;
        if (y.p == null) {
            root = x;
        } else if (y == y.p.left) {
            y.p.left = x;
        } else {
            y.p.right = x;
        }
        x.right = y;
        y.p = x;
    }

    private boolean colorOf(Node<Key> x) {
        return x == null ? BLACK : x.color;
    }

    private void setColor(Node<Key> x, boolean c) {
        if (x != null) {
            x.color = c;
        }
    }

    public void insertFixup(Node<Key> z) {
        assert z.color == RED;
        while (colorOf(z.p) == RED) {
            if (z.p == z.p.p.left) { // is z's parent a left child
                Node<Key> y = z.p.p.right; // y is z's uncle
                if (colorOf(y) == RED) { // are z's parent and uncle both red?
                    // case 1
                    setColor(z.p, BLACK);
                    setColor(y, BLACK);
                    setColor(z.p.p, RED);
                    z = z.p.p;
                } else {
                    if (z == z.p.right) {
                        // case 2
                        z = z.p;
                        rotateLeft(z);
                    }
                    // case 3
                    setColor(z.p, BLACK);
                    setColor(z.p.p, RED);
                    rotateRight(z.p.p);
                }
            } else { // same as above with "right" and "left" exchanged
                Node<Key> y = z.p.p.left;
                if (colorOf(y) == RED) {
                    setColor(z.p, BLACK);
                    setColor(y, BLACK);
                    setColor(z.p.p, RED);
                    z = z.p.p;
                } else {
                    if (z == z.p.left) {
                        z = z.p;
                        rotateRight(z);
                    }
                    setColor(z.p, BLACK);
                    setColor(z.p.p, RED);
                    rotateLeft(z.p.p);
                }
            }
        }
        root.color = BLACK;
    }

    public void put(Key key) {
        if (key == null) {
            throw new NullPointerException();
        }
        Node<Key> x = root; // node being compared with z
        if (x == null) {
            root = new Node<>(key);
            size = 1;
            root.color = BLACK;
            return;
        }
        Node<Key> y = null; // y will be parent of z
        int cmp;
        while (x != null) { // descend until reaching the sentinel
            y = x;
            cmp = key.compareTo(x.key);
            if (cmp < 0) {
                x = x.left;
            } else if (cmp > 0) {
                x = x.right;
            } else {
                return; // stop if duplicated keys are found
            }
        }
        Node<Key> z = new Node<>(key);
        z.p = y; // found the location: insert z with parent y (y != null)
        size++;
        cmp = key.compareTo(y.key);
        if (cmp < 0) {
            y.left = z;
        } else {
            y.right = z;
        }
        insertFixup(z);
    }

    private Node<Key> get(Key key) {
        if (key == null) {
            throw new NullPointerException();
        }
        Node<Key> x = root;
        while (x != null && key.compareTo(x.key) != 0) {
            int cmp = key.compareTo(x.key);
            if (cmp < 0) {
                x = x.left;
            } else {
                x = x.right;
            }
        }
        return x;
    }

    public boolean contains(Key key) {
        return get(key) != null;
    }

    private int height(Node<Key> x) {
        if (x == null) return -1;
        else return 1 + Math.max(height(x.left), height(x.right));
    }

    public int height() {
        return height(root);
    }
}
